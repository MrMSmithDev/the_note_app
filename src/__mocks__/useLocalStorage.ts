export const mockSaveNote = jest.fn();
export const mockRemoveNote = jest.fn();
export const mockRetrieveNotes = jest.fn().mockReturnValue([]);
export const mockLoadYear = jest.fn();
export const mockUpdateNote = jest.fn();

export const useLocalStorage = jest.fn(() => ({
  loadYear: mockLoadYear,
  saveNoteToLocal: mockSaveNote,
  removeNoteFromLocal: mockRemoveNote,
  retrieveNotesFromLocal: mockRetrieveNotes,
  updateNoteToLocal: mockUpdateNote,
}));
