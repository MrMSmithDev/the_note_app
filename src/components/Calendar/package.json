{
    "name": "calendar",
    "version": "1.0.0",
    "main": "Calendar.tsx",
    "dependencies": {
      "react": "^19.0.0"
    }
}