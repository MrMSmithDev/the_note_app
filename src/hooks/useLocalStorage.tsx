import React, { createContext, useContext, useState } from 'react';
import type LocalDataStruct from 'src/types/LocalDataStruct';
import Note from 'src/types/Note';

interface LocalStorageContextType {
  currentData: LocalDataStruct;
  saveNoteToLocal: (dateString: string, note: Note) => void;
  removeNoteFromLocal: (dateString: string, noteId: string) => void;
  loadYear: (year: number) => void;
  retrieveNotesFromLocal: (dateString: string) => Note[];
}

interface LocalStorageProviderProps {
  children: React.ReactNode;
}

const LocalStorageContext = createContext<LocalStorageContextType | undefined>(
  undefined
);

const LocalStorageProvider: React.FC<LocalStorageProviderProps> = ({
  children,
}) => {
  const [currentData, setCurrentData] = useState<LocalDataStruct>({});

  function loadYear(year: number) {
    const storedData = localStorage.getItem(year.toString());

    if (storedData) setCurrentData(JSON.parse(storedData));
    else setCurrentData({});
  }

  function saveNoteToLocal(dateString: string, note: Note) {
    const [yearStr, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    setCurrentData((prev) => {
      const updatedMonthData = {
        ...prev[month],
        [date]: [...(prev[month]?.[date] || []), note],
      };

      const updatedYearData = {
        ...prev,
        [month]: updatedMonthData,
      };

      localStorage.setItem(yearStr, JSON.stringify(updatedYearData));

      return updatedYearData;
    });
  }

  function removeNoteFromLocal(dateString: string, noteId: string) {
    const [yearStr, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    setCurrentData((prev) => {
      const updatedMonthData = {
        ...prev[month],
        [date]: prev[month]?.[date].filter((note) => note.id !== noteId),
      };

      const updatedYearData = {
        ...prev,
        [month]: updatedMonthData,
      };

      localStorage.setItem(yearStr, JSON.stringify(updatedYearData));

      return updatedYearData;
    });
  }

  function retrieveNotesFromLocal(dateString: string): Note[] {
    const [_, monthStr, dateStr] = dateString.split('-');
    const month = parseInt(monthStr, 10);
    const date = parseInt(dateStr, 10);

    return currentData[month]?.[date] || [];
  }

  return (
    <LocalStorageContext.Provider
      value={{
        currentData,
        saveNoteToLocal,
        removeNoteFromLocal,
        loadYear,
        retrieveNotesFromLocal,
      }}
    >
      {children}
    </LocalStorageContext.Provider>
  );
};

export function useLocalStorage(): LocalStorageContextType {
  const context = useContext(LocalStorageContext);
  if (!context) {
    throw new Error(
      'useLocalStorage must be used within a LocalStorageProvider'
    );
  }
  return context;
}

export default LocalStorageProvider;
