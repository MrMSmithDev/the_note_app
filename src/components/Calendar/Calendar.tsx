import React, { useEffect, useState } from 'react';
import Year from '@components/Year/';
import { motion, AnimatePresence } from 'framer-motion';
import { useLocalStorage } from '@hooks/useLocalStorage';
import ChevronIcon from '@components/icons/ChevronIcon';

const Calendar: React.FC = () => {
  const [fullYear, setFullYear] = useState<number>(new Date().getFullYear());
  const [animationDirection, setAnimationDirection] = useState<-1 | 1>(1);

  const { loadYear } = useLocalStorage();

  const token = null;

  function handleYearChange(e: React.MouseEvent<HTMLButtonElement>) {
    const valueChange = parseInt(e.currentTarget.value);
    setAnimationDirection(valueChange > 0 ? 1 : -1);
    setTimeout(() => {
      // Allows time for direction to set before animating change
      setFullYear((prev) => (prev += valueChange));
    }, 0);
  }

  useEffect(() => {
    if (token) {
    } else {
      loadYear(fullYear);
    }
  }, [fullYear]);

  return (
    <main className="h-[calc(100lvh-48px)] p-5">
      <nav className="flex justify-center px-10 items-center my-5 w-full">
        <button
          className="font-bold cursor-pointer disabled:opacity-50"
          onClick={handleYearChange}
          value={-1}
          disabled={fullYear <= 2010}
        >
          <span className="sr-only">previous</span>
          <ChevronIcon
            className="h-4 w-4 fill-[#0a0a0a] dark:fill-[#fff]"
            inverted={true}
          />
        </button>
        <AnimatePresence mode="wait">
          <motion.h1
            key={fullYear}
            initial={{ opacity: 0, x: animationDirection * 20 }}
            animate={{ opacity: 1, x: 0 }}
            exit={{ opacity: 0, x: -animationDirection * 20 }}
            transition={{ duration: 0.2 }}
            className="text-3xl font-bold font-kanit italics text-transparent bg-clip-text bg-linear-to-tr from-blue-400 to-blue-600 py-2 mx-10 w-50 text-center"
          >
            {fullYear}
          </motion.h1>
        </AnimatePresence>
        <button
          className="font-bold cursor-pointer disabled:opacity-50"
          onClick={handleYearChange}
          value={1}
          disabled={fullYear >= 2030}
        >
          <span className="sr-only">next</span>
          <ChevronIcon className="h-4 w-4 fill-[#0a0a0a] dark:fill-[#fff]" />
        </button>
      </nav>
      <div className="flex justify-center h-3/4 w-full">
        <AnimatePresence mode="wait">
          <motion.div
            key={fullYear}
            initial={{ opacity: 0, x: animationDirection * 20 }}
            animate={{ opacity: 1, x: 0 }}
            exit={{ opacity: 0, x: -animationDirection * 20 }}
            transition={{ duration: 0.2 }}
          >
            <Year year={fullYear} />
          </motion.div>
        </AnimatePresence>
      </div>
    </main>
  );
};

export default Calendar;
