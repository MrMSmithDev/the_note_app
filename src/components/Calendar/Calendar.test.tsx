import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Calendar from './Calendar';
import { mockLoadYear } from '@__mocks__/useLocalStorage';

// Mock useLocalStorage
jest.mock('@hooks/useLocalStorage', () =>
  require('@__mocks__/useLocalStorage')
);

// Mock Year component
jest.mock('@components/Year', () => require('@__mocks__/Year'));

describe('Calendar component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly to match snapshot', () => {
    const { asFragment } = render(<Calendar />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with the current year initially', () => {
    const currentYear = new Date().getFullYear();

    render(<Calendar />);
    expect(screen.getByText(currentYear.toString())).toBeInTheDocument();
  });

  it('calls loadYear with the correct year on mount', () => {
    // const { loadYear } = useLocalStorage();
    const currentYear = new Date().getFullYear();

    render(<Calendar />);

    expect(mockLoadYear).toHaveBeenCalledWith(currentYear);
  });

  it('increments the year when the next button is clicked', async () => {
    render(<Calendar />);

    const nextButton = screen.getByRole('button', { name: /next/i });
    const currentYear = new Date().getFullYear();

    fireEvent.click(nextButton);

    expect(
      await screen.findByText((currentYear + 1).toString())
    ).toBeInTheDocument();
  });

  it('decrements the year when the previous button is clicked', async () => {
    render(<Calendar />);

    const prevButton = screen.getByRole('button', { name: /previous/i });
    const currentYear = new Date().getFullYear();

    fireEvent.click(prevButton);

    expect(
      await screen.findByText((currentYear - 1).toString())
    ).toBeInTheDocument();
  });

  it('disables the previous button at the lower year limit (2010)', async () => {
    render(<Calendar />);

    const prevButton = screen.getByRole('button', { name: /previous/i });

    for (let i = new Date().getFullYear(); i > 2009; i--) {
      fireEvent.click(prevButton);
    }

    await waitFor(() => expect(prevButton).toBeDisabled());
  });

  it('disables the next button at the upper year limit (2030)', async () => {
    render(<Calendar />);

    const nextButton = screen.getByRole('button', { name: /next/i });

    for (let i = new Date().getFullYear(); i < 2030; i++) {
      fireEvent.click(nextButton);
    }

    await waitFor(() => expect(nextButton).toBeDisabled());
  });
});
