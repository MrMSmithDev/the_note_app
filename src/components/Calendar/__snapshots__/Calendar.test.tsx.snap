// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component renders correctly to match snapshot 1`] = `
<DocumentFragment>
  <main
    class="h-[calc(100lvh-48px)] p-5"
  >
    <nav
      class="flex justify-center px-10 items-center my-5 w-full"
    >
      <button
        class="font-bold cursor-pointer disabled:opacity-50"
        value="-1"
      >
        <span
          class="sr-only"
        >
          previous
        </span>
        <svg
          class="h-4 w-4 fill-[#0a0a0a] dark:fill-[#fff] rotate-180"
          height="200px"
          viewBox="0 0 32 32"
          width="200px"
        >
          <g
            stroke-width="0"
          />
          <g
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <g>
            <g>
              <path
                d="M12.5,28c-1,0-1.8-0.4-2.5-1c-1.4-1.4-1.4-3.6,0-5l5.7-5.8c0.1-0.1,0.1-0.2,0-0.3L10,10.1c-1.4-1.4-1.4-3.6,0-5 c0.7-0.7,1.6-1,2.5-1c0,0,0,0,0,0c1,0,1.8,0.4,2.5,1l9.3,9.4c0.9,0.9,0.9,2.3,0,3.1l-9.3,9.4C14.4,27.6,13.5,28,12.5,28 C12.5,28,12.5,28,12.5,28z"
              />
               
            </g>
          </g>
        </svg>
      </button>
      <h1
        class="text-3xl font-bold font-kanit italics text-transparent bg-clip-text bg-linear-to-tr from-blue-400 to-blue-600 py-2 mx-10 w-50 text-center"
        style="opacity: 0; transform: translateX(20px);"
      >
        2025
      </h1>
      <button
        class="font-bold cursor-pointer disabled:opacity-50"
        value="1"
      >
        <span
          class="sr-only"
        >
          next
        </span>
        <svg
          class="h-4 w-4 fill-[#0a0a0a] dark:fill-[#fff] "
          height="200px"
          viewBox="0 0 32 32"
          width="200px"
        >
          <g
            stroke-width="0"
          />
          <g
            stroke-linecap="round"
            stroke-linejoin="round"
          />
          <g>
            <g>
              <path
                d="M12.5,28c-1,0-1.8-0.4-2.5-1c-1.4-1.4-1.4-3.6,0-5l5.7-5.8c0.1-0.1,0.1-0.2,0-0.3L10,10.1c-1.4-1.4-1.4-3.6,0-5 c0.7-0.7,1.6-1,2.5-1c0,0,0,0,0,0c1,0,1.8,0.4,2.5,1l9.3,9.4c0.9,0.9,0.9,2.3,0,3.1l-9.3,9.4C14.4,27.6,13.5,28,12.5,28 C12.5,28,12.5,28,12.5,28z"
              />
               
            </g>
          </g>
        </svg>
      </button>
    </nav>
    <div
      class="flex justify-center h-3/4 w-full"
    >
      <div
        style="opacity: 0; transform: translateX(20px);"
      >
        <div
          data-testid="mock-year"
        >
          mock year
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
